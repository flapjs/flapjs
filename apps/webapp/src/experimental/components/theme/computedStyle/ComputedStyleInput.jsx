import React from 'react';
import PropTypes from 'prop-types';
import StyleInput from '../sourceStyle/StyleInput.jsx';

import { getElementFromRef } from '@flapjs/deprecated/util/ElementHelper.js';
import { TinyColor } from '@ctrl/tinycolor';

/**
 * A React component that can compute and control a style
 * property of one or another DOM element.
 */
class ComputedStyleInput extends React.Component
{
    constructor(props)
    {
        super(props);

        this.self = React.createRef();

        this.onComputeChange = this.onComputeChange.bind(this);
    }
    
    /** @override */
    componentDidMount()
    {
        const computeElement = getElementFromRef(this.props.compute);
        if (computeElement)
        {
            computeElement.addEventListener('change', this.onComputeChange);
        }
    }
    
    /** @override */
    componentWillUnmount()
    {
        const computeElement = getElementFromRef(this.props.compute);
        if (computeElement)
        {
            computeElement.removeEventListener('change', this.onComputeChange);
        }
    }

    onComputeChange(e)
    {
        const computeFunction = this.props.computeFunction;
        if (typeof computeFunction === 'function')
        {
            computeFunction.call(null, e.value);
        }
        else
        {
            switch(computeFunction)
            {
                case 'darken':
                    this.self.current.setValue(new TinyColor(e.value).darken(20).toHexString());
                    break;
                case 'lighten':
                    this.self.current.setValue(new TinyColor(e.value).brighten(20).toHexString());
                    break;
                case 'invert':
                    this.self.current.setValue(new TinyColor(e.value).complement().toHexString());
                    break;
                case 'copy':
                    this.self.current.setValue(e.value);
                    break;
                default:
                    throw new Error(`Unknown compute function '${computeFunction}.'`);
            }
        }
    }

    /** @override */
    render()
    {
        const props = this.props;
        
        return (
            <StyleInput
                ref={this.self}
                className={props.className}
                source={props.source}
                name={props.name}
                type={props.type}
                disabled={props.disabled}
                onChange={props.onChange} />
        );
    }
}

ComputedStyleInput.propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    source: PropTypes.oneOfType([
        PropTypes.shape({
            current: PropTypes.instanceOf(Element)
        }),
        PropTypes.func,
    ]).isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.oneOf([
        'color',
        'text',
        'number',
    ]),
    onChange: PropTypes.func,
    compute: PropTypes.oneOfType([
        PropTypes.shape({
            current: PropTypes.instanceOf(StyleInput)
        }),
        PropTypes.func,
    ]).isRequired,
    computeFunction: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOf([
            'darken',
            'lighten',
            'invert',
            'copy',
        ])
    ]),
};
ComputedStyleInput.defaultProps = {
    disabled: false,
    type: 'color',
    computeFunction: 'copy',
};

export default ComputedStyleInput;
