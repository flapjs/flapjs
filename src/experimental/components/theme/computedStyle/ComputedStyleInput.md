# ComputedStyleInput

## Motivation
Cause I love all of you. <3

## Work Log
- Created ComputedStyleInput component.