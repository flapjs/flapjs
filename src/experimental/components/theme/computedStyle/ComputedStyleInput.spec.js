import React from 'react';
import { shallow } from 'enzyme';

import ComputedStyleInput from './ComputedStyleInput.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(
        <ComputedStyleInput
            source={{ current: null }}
            name=""
            compute={{ current: null }} />
    );
});

describe('<ComputedStyleInput/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<ComputedStyleInput/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<ComputedStyleInput/> lifecycle', () =>
{
    /** ...later. */
});
