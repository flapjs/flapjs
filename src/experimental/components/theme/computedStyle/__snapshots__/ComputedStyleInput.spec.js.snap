// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComputedStyleInput/> rendering renders correctly 1`] = `
<StyleInput
  disabled={false}
  name=""
  source={
    Object {
      "current": null,
    }
  }
  type="color"
/>
`;
